Masking

mike-spa opened the pull request:
Resolves: #26135 
Resolves: #11570
Resolves: #22646

![image](https://github.com/user-attachments/assets/5ee31c3b-116c-4e56-993a-edb70c158b38)


This PR implements a general infrastructure for masking any item drawn on the score. It should be noted that this is not achieved by drawing rectangles above/below the items (which is a hack that causes all sorts of problems, such as the ones described in #11570, and anyway is not a general solution), but rather by natively _not drawing_ in the masked region. 

Currently, this new system is applied to

- **Barline VS Text**. Any text-based item which crosses a barline will now cause the barline to be automatically masked. This forms part of the solution to #26135.
- **Fret numbers VS String Lines in TABs**. This resolves #11570. 

An option to control the first point has been added to the Barline style page.
![image](https://github.com/user-attachments/assets/fbe406ee-48d0-40ba-ae96-708b754f5ae5)

Additionally as part of the solution to #26135, we've introduced a new option for Lyrics to _avoid_ Barlines. This probably deserved a PR of its own cause it's not stricly related to masking, but the two things came up together so here they are. Obviously, this option is only relevant for the case of barlines across the staves. 
The general option is in the Lyrics style page
![image](https://github.com/user-attachments/assets/b255ccee-bf6d-4031-b3e9-4d8e15d5fa8a)
and the local property override is in the Lyrics property page.
![image](https://github.com/user-attachments/assets/d4fc452b-e970-4ba3-8b76-2b27d3413389)
This is similar to what we've already done with Dynamics.

cbjeukendrup commented on src/framework/draw/internal/qpainterprovider.cpp:
```diff
@@ -357,6 +357,26 @@ void QPainterProvider::setClipRect(const RectF& rect)
+
+    QPainterPath mask = background.subtracted(exclusionRegion);
+
+    m_painter->setClipPath(mask);
```
Do you know how this affects performance? I think I heard at some point that clipping is relatively expensive, and I wonder if the size of the clipping area contributes to that or not. Perhaps there's no noticeable difference in our case, but it would be good to do a bit of testing, for example by looking at how smoothly the score moves when scrolling/dragging/zooming quickly.

mike-spa replied:
Rectangular clipping like this is ["guaranteed" to be efficient](https://doc.qt.io/qt-6/qpainter.html#clipping:~:text=Rectangular%20clipping%20with%20simple%20transformations%20and%20intersect%20clip.). Also, intuition tells me that the size of the clipping area may only affect performance for rasterized graphics, not vector drawing.

cbjeukendrup reviewed (commented):
Here's a quick review of especially the files from `framework` and below.

In general, I think this is the right solution. It's not very intuitive in my opinion; more ideal would be an item's own `draw` function would clear away the area behind that item. But that would be quite difficult to implement, and in the implementation I imagined, it would be impossible / even more complicated to erase only _some_ elements from the background; which would be needed for example when we want to mask ties that cross time signatures, because then only the tie should be cut away, and not the staff lines. Also, it would depend too much on z order. So it seems your solution is the most viable one. 

With any solution to this problem, there is the risk of performance impact, so we should do some testing for that.

Another thing to make sure is that it also works correctly with SVG export. Not sure if the SVG exporter supports clip paths or clipping at all yet.

mike-spa commented:
> In general, I think this is the right solution. It's not very intuitive in my opinion; more ideal would be an item's own `draw` function would clear away the area behind that item. But that would be quite difficult to implement, and in the implementation I imagined, it would be impossible / even more complicated to erase only _some_ elements from the background; which would be needed for example when we want to mask ties that cross time signatures, because then only the tie should be cut away, and not the staff lines. Also, it would depend too much on z order. So it seems your solution is the most viable one.

@cbjeukendrup yep, all correct. I should add that, because QPainterPath supports subtraction, my preferred solution would have been to clip the path that we're _passing_ to the painter, rather than clip the painter itself. However, because we use our own implementation of PainterPath (incidentally, do you know the reason why we did this? why not just hide QPainterPath behind an interface and rely on that?) we don't have that functionality available. And, as you said, implementing general clipping ourselves would be _extremely_ difficult. So the next best thing we can do is clip the painter.

> With any solution to this problem, there is the risk of performance impact, so we should do some testing for that.

I am very confident that the drawing performance is unaffected by these changes. Both because only a tiny fraction of all the drawn objects will be actually clipped. But also because Qt "guarantees" that simple rectangular clipping is [safe to use in performance-critical scenarios](https://doc.qt.io/qt-6/qpainter.html#clipping:~:text=Rectangular%20clipping%20with%20simple%20transformations%20and%20intersect%20clip.). For us, the expensive task is _computing_ the masks, i.e. the MaskLayout functions. I'll do a quick check on that, but I'm sure we're fine.

> Another thing to make sure is that it also works correctly with SVG export. Not sure if the SVG exporter supports clip paths or clipping at all yet.

That is a very good observation and in fact I've just tested it and it _doesn't_ work with SVG. We'll just have to document that, I guess.

mike-spa commented on src/framework/draw/painter.h:
```diff
@@ -169,6 +169,7 @@ class Painter
     bool hasClipping() const;
 
     void setClipRect(const RectF& rect);
+    void setMask(const mu::engraving::Shape& mask);
```
Yeah I was also unsure about this, cause I too didn't totally like the idea of passing a Shape down. I am going to refactor this and pass down a vector of rectangles instead (which basically is what a Shape is, anyway).

igorkorsukov replied:
now the framework is already used in other products where there is no engraving (in Audacity), so it is impossible to use something from outside in the framework   

To solve such problems, we can introduce a new type at the framework level and use it or convert to it.
Or in this case, Shape is a vector of RectF.

cbjeukendrup commented:
> However, because we use our own implementation of PainterPath (incidentally, do you know the reason why we did this? why not just hide QPainterPath behind an interface and rely on that?) we don't have that functionality available.

This was @igorkorsukov's choice; it's a logical choice, because for value types, like PainterPath, you can't really create interfaces. That's only done for types that are only passed by reference/pointers.

bkunda reviewed (changes requested):
Love it!

Only I can't seem to get the lyrics masking to work (even after a factory reset, and despite the new checkbox option in Style being enabled).

It does work for expression text, though:

https://github.com/user-attachments/assets/71d2cf12-2faa-447c-b7af-52ddf0640533

Also a bug seems may have been introduced in this PR, which does not occur in the latest nightly: the note input cursor somehow becomes very opaque when the mouse cursor is moved:

https://github.com/user-attachments/assets/ee31d066-0686-4dff-930d-3f396cc102d3

Tested only on macOS so far, so I half wonder whether these issues may be OS-specific.

igorkorsukov commented on src/engraving/infrastructure/shape.cpp:
```diff
@@ -144,6 +144,24 @@ Shape Shape::adjusted(double xp1, double yp1, double xp2, double yp2) const
+    return *this;
+}
+
+Shape Shape::padded(double p)
```
this method can be const

igorkorsukov commented:
> In general, I think this is the right solution. It's not very intuitive in my opinion; more ideal would be an item's own draw function would clear away the area behind that item. But that would be quite difficult to implement, and in the implementation I imagined, it would be impossible / even more complicated to erase only some elements from the background; which would be needed for example when we want to mask ties that cross time signatures, because then only the tie should be cut away, and not the staff lines. Also, it would depend too much on z order. So it seems your solution is the most viable one.

@mike-spa @cbjeukendrup 
Just for understanding.
Why isn't it easier to solve this problem by simply defining the drawing order, first drawing the backgrounds elements, barlines, etc. below. And then drawing the overlying elements on top with their filled area with a color matching the background color. This would work universally for any elements, and the whole difficulty is in defining and drawing the elements in a given order, I think it's not difficult.

igorkorsukov commented:
> > However, because we use our own implementation of PainterPath (incidentally, do you know the reason why we did this? why not just hide QPainterPath behind an interface and rely on that?) we don't have that functionality available.
> 
> This was @igorkorsukov's choice; it's a logical choice, because for value types, like PainterPath, you can't really create interfaces. That's only done for types that are only passed by reference/pointers.

Besides, if it’s difficult for us to make some function without Qt, then it will be difficult for others to do it too... And our neighboring project doesn’t have Qt.

igorkorsukov commented:
> And, as you said, implementing general clipping ourselves would be extremely difficult. So the next best thing we can do is clip the painter. 

It would be good to find out whether our neighboring internal project can implement such clipping.

mike-spa commented:
> Only I can't seem to get the lyrics masking to work (even after a factory reset, and despite the new checkbox option in Style being enabled).

@bkunda that's because we've introduced a further option in the lyrics style page which by default is set to avoid barline collisions. If you deactivate that style option, or deactivate the individual property override for that single lyrics, it will let the collision happen and you'll see the masking.

> Also a bug seems may have been introduced in this PR, which does not occur in the latest nightly: the note input cursor somehow becomes very opaque when the mouse cursor is moved:

Fixed now, thank you :)

mike-spa commented:
> Why isn't it easier to solve this problem by simply defining the drawing order, first drawing the backgrounds elements, barlines, etc. below. And then drawing the overlying elements on top with their filled area with a color matching the background color.

Because it isn't a very general solution @igorkorsukov. The main problems are:
1. The background is not necessarily a solid color.
![image](https://github.com/user-attachments/assets/fc504de4-5a4d-47a9-9bf9-81ef56bf4ac7)
2. The background can be transparent, in which this doesn't work anymore.
3. There can be situations where it's not possible to find a suitable drawing order. In this example, we want the key signature and time signature to mask the ties but _not_ the barlines, but at the same time we need all of them to stay _in front_ of the barlines. We can't achieve this with a rectangle. ![image](https://github.com/user-attachments/assets/d99716a2-31d4-433f-8477-8a99a4413496)

This solution using the clipPath from QPainter is very simple for us so I think we should try to go with it. I am sure that masking is possible also in SVG, because if I convert the output PDF from Musescore into SVG (using an external tool), the barlines are masked correctly, so we probably just need to do a bit of research and find out how to do it. 
![image](https://github.com/user-attachments/assets/c83a1195-576b-4b3b-abc9-14aba9c798b7)
Same for our neighbouring projects without Qt. Any vector-drawing tool should be capable of doing this I think. 

For now, as this is a hyper-specialized engraving feature, I don't think it's a problem if the masking isn't exported to SVG or other non-Qt viewers.

avvvvve commented:
Love to see it! Live masking while dragging was especially exciting to see in action:

https://github.com/user-attachments/assets/c2e9729c-7db7-4574-a424-73d6190a7b87

My main feedback here was already mentioned (SVG exported doesn't work), but if there's a workaround via PDF to SVG conversion, then I think we're okay.

Approving from my end, but this could definitely use some deeper QA testing.

-----

If we're not planning on supporting proper SVG export for 4.5, here's a case we should be prepared to respond to with the PDF solution if we start hearing from users that we broke it:

The old SVG export solution (i.e. draw a white box behind the element) wasn't great either, but at least if you had the boxes there, you could use them as guides to manually clip the path in your vector editing program of choice.

<img width="780" alt="image" src="https://github.com/user-attachments/assets/46bea13e-b6bc-47e1-b9ae-9e669ae6e5af" />

Now, nothing gets exported beneath the masking elements, so if you wanted to cut bits of the line out, you won't have an easy, pre-placed guide to do so.

<img width="644" alt="image" src="https://github.com/user-attachments/assets/12f527eb-52fd-4000-9dae-6d627b1f80fb" />

bkunda reviewed (approved):
Very happy with all this functionality, although I'd like a design review of the new options in the Style dialog ("lyrics" and "barlines") pages. It's not ideal that you have to interact with checkboxes in both places to get lyrics to pass through barlines with masking on.

But this is a minor detail. The affordances provided by this PR are truly excellent.

mike-spa commented:
Update re SVG export @avvvvve @cbjeukendrup @igorkorsukov. As it turns out, the SVG standard does have an exact equivalent, which also happens to be called `clipPath` (see [here](https://www.w3schools.com/graphics/svg_clipping_masking.asp) for details), so it was quite easy to implement it (see the last commit). Now also our SVG export works perfectly with the new masking system :) 

@cbjeukendrup it'd be great if you could give this a final round of code review so we can merge it soon.

igorkorsukov commented on src/framework/draw/ipaintprovider.h:
```diff
@@ -33,6 +33,8 @@
 #include "types/transform.h"
 #include "types/painterpath.h"
 
+#include "../../engraving/infrastructure/shape.h"
```
It would be nice to delete

Tantacrul commented:
To confirm, the masking options are off by default for older scores, on by default for new ones. (Although with Lyrics, you obviously need to switch off the 'avoid bar lines' first)?

Agree with @bkunda that it would probably be good to fix up those masking settings to try and find a way to make them a bit more intuitive to locate. Worth a checkbox in 'Properties' for relevant items? Not sure.

Tantacrul commented:
When testing a little bit again, I've a few questions:

---
When I export as a PDF and open in Illustrator, each of the masks are simply white boxes overlaying the background. My understanding was that this wasn't the desired approach, so I'm wondering if it's an export issue or something else I'm missing, @mike-spa ?

What it looks like
<img width="203" alt="image" src="https://github.com/user-attachments/assets/9056302f-e384-4744-a72c-e6b959d5cba4" />

Selecting the box
<img width="397" alt="image" src="https://github.com/user-attachments/assets/c111d499-322e-483d-9b89-525954ee6902" />

Deleting the box
<img width="370" alt="image" src="https://github.com/user-attachments/assets/d2f962df-7b05-47de-bba2-a727c064cb24" />

---

Also, I'm not sure if this is desired but it doesn't work on the final barline (although it does work on when changing a regular barline to look like a final barline.

<img width="475" alt="image" src="https://github.com/user-attachments/assets/c5c317a3-0d9b-4e3c-b132-8ab36c363ac9" />

Tantacrul reviewed (changes requested):
Curious to understand what the desired outcome is for exported SVGs and also PDF export.

mike-spa commented:
> When I export as a PDF and open in Illustrator, each of the masks are simply white boxes overlaying the background.

@Tantacrul I think that is mostly likely due to Illustrator itself and how it _interprets_ the masks. I don't have Illustrator or other PDF editors, but if I convert Musescore's output PDF in SVG (using a random online tool), my SVG gets perfecly masked lines, not white rectangles, which tells me that the correct masking information _is_ in the PDF. In any case, for PDF export we use 100% Qt so I wouldn't have any control over it anyway.
![image](https://github.com/user-attachments/assets/f67cfe53-10a4-4746-a459-b99c663908d2)

> Also, ditto @avvvvve's observation that cutaways don't work for SVG export, whether or not you choose a transparent BG.

I think you're testing the older build, pls try out the latest one. I have now implemented masks in the SVG export (which, unlike PDF, we do ourselves so I can control it) and my output SVG looks like this. I'm using Inkscape to open all the SVGs btw.
![image](https://github.com/user-attachments/assets/3a2ad97c-6558-4c8f-8f40-e92de566b848)

> Also, I'm not sure if this is desired but it doesn't work on the final barline

It's not about it being "final" but being at the end of the system, and that's intentional. I don't think the start- and end barline of a system should ever be masked (not least because it means that you're placing stuff in the page margins). Though @oktophonie should have last word on this really.

> Lastly, would we consider doing this for hairpins also?

Well, I'd say certainly not now cause it would need yet another style/property option. As we develop this further (the really big one I'm looking forward to is masking ties...) it may be worth, as you and @bkunda mentioned, to think about something better, perhaps a style page specific to masking. But in general I think most of the time barlines aren't masked for other lines (though again @oktophonie territory).

Tantacrul commented:
@mike-spa 

For PDF, the only thing that concerns me are cases like the one you mentioned to Igor above:

> There can be situations where it's not possible to find a suitable drawing order. In this example, we want the key signature and time signature to mask the ties but not the barlines, but at the same time we need all of them to stay in front of the barlines. We can't achieve this with a rectangle.

![image](https://github.com/user-attachments/assets/9ccc0423-16d1-4b22-8dae-434056c1fe7a)

---

If PDF's render white boxes for those situations, we might get into trouble. However, I can't know for sure until the thing is implemented. It's possible the PDF might be more 'clever' in how it distributes masks?

Tantacrul commented:
> I think you're testing the older build, pls try out the latest one. 


Excuse my ludidte-ness, but I'm trying out the build associated with this conversation in 'Checks' and SVG export is still showing the same problems as I mentioned above.

avvvvve commented:
@mike-spa @Tantacrul In the build I downloaded today, I'm seeing that SVG export is masking things as expected on macOS!

-----

I think there are some improvements to be made in terms of _how_ masking is done, but not sure if they're easily solvable. See video:

https://github.com/user-attachments/assets/63f16197-5e77-4e54-87c2-fffe2db9abe1

-----

Also, if I export a PDF and open it directly in illustrator (as a PDF), it looks correct:
<img width="300" alt="image" src="https://github.com/user-attachments/assets/79d53f29-004f-473a-9c07-07b1a1ab58ff" />

However, if I then export the open PDF file in Illustrator to SVG, then open the resulting SVG in any editor, the clipping is broken (this isn't really an issue though if our native SVG export works correctly):
<img width="300" alt="image" src="https://github.com/user-attachments/assets/633cf50e-8691-43c4-ab85-37c75e13d877" />

avvvvve commented:
<img width="358" alt="image" src="https://github.com/user-attachments/assets/a702d544-b40c-4730-84d6-08f5d672c141" />

Small request: In Styles, we move "Mask barlines when intersecting text" next to the other checkboxes, just beneath "Scale barlines to staff size"?
